Scripts/Runtime/Util/Jobs/ParallelCollectionUtil.cs

```diff
@@ -0,0 +1,103 @@
+        }
+
+        /// <summary>
+        /// Returns the correct index for the collection based on the <paramref name="nativeThreadIndex"/> passed in.
+        /// </summary>
+        /// <remarks>
+        /// This function assumes that the collection being used was sized appropriately via
+        /// <see cref="ParallelCollectionUtil.CollectionSizeForMaxThreads"/>. Larger sized collections will still work
+        /// but the intended usage is to create a small tightly packed collection sized for the specific hardware the
+        /// program is running on. Smaller sized collections will error sporadically as the scheduler assigns jobs to
+        /// out of range thread indexes.
+        ///
+        /// This function also assumes that you are never running your job on the main thread via
+        /// <see cref="IJobExtensions.Run"/>. In that case the thread index will be 0 and this function will
+        /// return -1 which will cause an error.
+        ///
+        /// When scheduling your job, Unity will place your job on one of the available worker threads in which case
+        /// the native thread index you receive will be from 1 to <see cref="JobsUtility.JobWorkerMaximumCount"/>.
+        /// There are two special cases.
+        /// 1. The scheduler may place your job on the main thread. In this case the native thread index will be
+        /// <see cref="JobsUtility.JobWorkerMaximumCount"/> + 2.
+        /// 2. The scheduler may place your job on a profiler thread in the editor. In this case the native thread index
+        /// will also be <see cref="JobsUtility.JobWorkerMaximumCount"/> + 2.
+        /// It is unknown why this is. 
+        ///
+        /// Our goal is to have a tightly packed collection so in the example of an 8 core machine with
+        /// <see cref="JobsUtility.JobWorkerMaximumCount"/> equal to 15, we would have the following mapping.
+        ///
+        /// thread 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17 
+        /// index  0, 1, 2, 3, 4, 5, 6, 7, 8, 9,  10, 11, 12, 13, 14, 15
+        ///
+        /// We have a tightly packed collection with index 0 through 15 for a total of 16 buckets.
+        /// Thread indexes map directly without having to remember the special rules.
+        /// </remarks>
+        /// <param name="nativeThreadIndex">
+        /// The native thread index to figure out the collection index from. Must be greater than 0.
+        /// </param>
+        /// <returns>The collection index to use</returns>
+        public static int CollectionIndexForThread(int nativeThreadIndex)
+        {
+            // JobsUtility.JobWorkerMaximumCount = 15
+            // index  0, 1, 2, 3, 4, 5, 6, 7, 8, 9,  10, 11, 12, 13, 14, 15
+            // thread 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17
+            return (nativeThreadIndex <= JOB_WORKER_MAXIMUM_COUNT)
+                ? nativeThreadIndex - 1
+                : JOB_WORKER_MAXIMUM_COUNT;
```

Comment:
```suggestion
            return math.min(nativeThreadIndex - 1, JOB_WORKER_MAXIMUM_COUNT);
```

Would this work better?

Reply:
Yes good call